src/pool.c

```diff
@@ -23,15 +23,24 @@ static void *pool_alloc_page(git_pool *pool, uint32_t size);
 
 uint32_t git_pool__system_page_size(void)
 {
+#ifndef GIT_THREADS
 	static uint32_t size = 0;
 
 	if (!size) {
+#else
+    uint32_t size = 0;
+    do {
+#endif
 		size_t page_size;
 		if (git__page_size(&page_size) < 0)
 			page_size = 4096;
 		/* allow space for malloc overhead */
 		size = page_size - (2 * sizeof(void *)) - sizeof(git_pool_page);
+#ifndef GIT_THREADS
 	}
+#else
+    } while (0);
+#endif
```

Comment:
Well, this effectively removes the cache altogether on most platforms. On Linux-based systems this is basically not an issue, as `sysconf` will usually just return a static value without any syscalls at all. But on Windows this might be a different story, as we call `GetSystemInfo` there, which might be a bit more involved. And as we're using pools especially in performance-critical paths, this is a no-go.

In the end, the reported threading issue is really a non issue. There's only two states: the value either is uninitialized or it is initialized to a value that is never going to change during execution. So what happens in case two threads race? The worst thing that can happen is that the value gets initialized to the same value twice, which really is not a problem at all. And as stores are guaranteed to be atomic, another thread will only ever see either the correct value or an uninitialized value.

I agree that fixing issues reported by TSAN is nice to make TSAN itself more useful. But I'd argue that pessimizing a potentially performance critical path in order to fix a reported issue that cannot even have any negative consequences is the wrong thing to do.

Reply:
There's no such thing as a benign data race. I don't know where you're getting the idea that stores are guaranteed to be atomic. If one stops assuming the happy path of a bog-standard x86 running some archaic Linux, then it becomes obvious that a problem like this must be taken seriously. The read of `size` in `if (!size)` can happen at any point during someone else's write to it.

Is there instrumentation data indicating that a cache is needed? Your uncertainty implies that there isn't. Every use I can find of a modern system that uses the page size in a calculation, including many platform allocators, use the `sysconf` on demand.

Reply:
>  The read of size in if (!size) can happen at any point during someone else's write to it.

I agree with you that if we’re running on a machine that can’t atomically store a 32 bit number then this would be problematic.  Can you give me a concrete example of such a machine?